primitives/payment-handler/src/fee_handler.rs

```diff
@@ -46,73 +47,104 @@ impl<AccountId, Balance> Payment<AccountId, Balance> {
+pub type AffiliateFeeDistribution<AccountId, Balance, MaxDistribution> =
+	BoundedVec<PaymentFee<AccountId, Balance>, MaxDistribution>;
+pub type TournamentFeeDistribution<AccountId, Balance> = PaymentFee<AccountId, Balance>;
+
+pub struct AssetGameFeeHandler<
+	AccountId,
+	PaymentAssets,
+	WithdrawPaymentAssets,
+	Affiliate,
+	AffiliateMaxDistribution,
+	Tournament,
+> {
+	_phantom: PhantomData<(
+		AccountId,
+		PaymentAssets,
+		WithdrawPaymentAssets,
+		Affiliate,
+		AffiliateMaxDistribution,
+		Tournament,
+	)>,
 }
 
-impl<AccountId, Assets, W, Affiliate, Tournament> FeeHandler
-	for AssetGameFeeHandler<AccountId, Assets, W, Affiliate, Tournament>
-where
-	// This is satisfied by the `pallet-assets`, `pallet-asset-conversion` and the `NativeAndAssets`
-	// struct.
-	Assets: fungibles::Inspect<AccountId, Balance = W::Balance, AssetId = W::AssetId>
-		+ fungibles::Balanced<AccountId>,
-
-	W: WithdrawCredit<
+impl<AccountId, PaymentAssets, WPA, Affiliate, AffiliateMaxDistribution, Tournament> FeeHandler
+	for AssetGameFeeHandler<
+		AccountId,
+		PaymentAssets,
+		WPA,
+		Affiliate,
+		AffiliateMaxDistribution,
+		Tournament,
+	> where
+	// This is satisfied by the `pallet-assets`, `pallet-asset-conversion` and the
+	// `NativeAndAssets` struct.
+	PaymentAssets:
+		fungibles::Inspect<AccountId, Balance = WPA::Balance> + fungibles::Balanced<AccountId>,
+	WPA: WithdrawCredit<AccountId = AccountId, Assets = PaymentAssets>,
+	WPA::Credit: IntoFungiblesCredit<AccountId, PaymentAssets>,
+
+	Affiliate: DistributeFee<
 		AccountId = AccountId,
-		Assets = Assets,
-		Credit = fungibles::Credit<AccountId, Assets>,
+		Balance = WPA::Balance,
+		FeeDistribution = AffiliateFeeDistribution<
+			AccountId,
+			WPA::Balance,
+			AffiliateMaxDistribution,
+		>,
+	>,
+	Tournament: DistributeFee<
+		AccountId = AccountId,
+		Balance = WPA::Balance,
+		FeeDistribution = TournamentFeeDistribution<AccountId, WPA::Balance>,
 	>,
-
-	Affiliate: DistributeFee<AccountId = W::AccountId, Balance = W::Balance>,
-	Tournament:
-		DistributeFee<AccountId = AccountId, Balance = W::Balance, MaxDistributions = ConstU32<1>>,
 {
 	type AccountId = AccountId;
-	type AssetId = W::AssetId;
-	type Balance = W::Balance;
+	type Payment = WPA::AssetId;
+	type Balance = WPA::Balance;
 	type AffiliateFeeIdentifier = Affiliate::FeeIdentifier;
 	type TournamentFeeIdentifier = Tournament::FeeIdentifier;
 
 	fn withdraw_and_pay_fees(
 		payer: &Self::AccountId,
-		payment_asset: Self::AssetId,
+		payment: Self::Payment,
 		base_fee: Self::Balance,
 		tournament_id: &Self::TournamentFeeIdentifier,
 		affiliate_id: &Self::AffiliateFeeIdentifier,
 		treasury_pot: &Self::AccountId,
 	) -> Result<(), DispatchError> {
 		// The credit may be in any asset as implemented by `WithdrawAsset`.
-		let fee_credit = W::withdraw_credit(payer, payment_asset, base_fee)?;
+		let fee_credit = WPA::withdraw_credit(payer, payment.clone(), base_fee)?.into_credit();
```

Comment:
There is now a very valid reason that this is 0. So we should do a check and do an early return in this case.

As mentioned in `withdraw_credit`, I don't think that we need the `into_credit` implementations if we internally convert it to `Credit.zero()`.

However, I just notice now. We could still return an option here, and then just early return if it is `None`. Either way, I don't think we need the `into_credit` boilerplate`.